commcare_connect/form_receiver/processor.py

```diff
@@ -271,59 +274,77 @@ def process_deliver_unit(user, xform: XForm, app: CommCareApp, opportunity: Oppo
-            form_json=xform.raw_form,
-            location=xform.metadata.location,
+    user_visit = UserVisit(
+        opportunity=opportunity,
+        user=user,
+        opportunity_access=access,
+        deliver_unit=deliver_unit,
+        entity_id=entity_id,
+        entity_name=entity_name,
+        visit_date=xform.metadata.timeStart,
+        xform_id=xform.id,
+        app_build_id=xform.build_id,
+        app_build_version=xform.metadata.app_build_version,
+        form_json=xform.raw_form,
+        location=xform.metadata.location,
+    )
+    today = datetime.date.today()
+    paymentunit_startdate = payment_unit.start_date if payment_unit else None
+    if opportunity.start_date > today or (paymentunit_startdate and paymentunit_startdate > today):
+        completed_work = None
+        user_visit.status = VisitValidationStatus.trial
+    else:
+        completed_work = (
+            CompletedWork.objects.filter(opportunity_access=access, entity_id=entity_id, payment_unit=payment_unit)
+            .annotate(rank=Window(Rank(), order_by="date_created"))
+            .filter(rank=counts["entity"] + 1)
+            .order_by("rank")
+            .last()
         )
-        completed_work_needs_save = False
-        today = datetime.date.today()
-        paymentunit_startdate = payment_unit.start_date if payment_unit else None
-        if opportunity.start_date > today or (paymentunit_startdate and paymentunit_startdate > today):
-            completed_work = None
-            user_visit.status = VisitValidationStatus.trial
-        else:
-            completed_work, _ = CompletedWork.objects.get_or_create(
-                opportunity_access=access,
-                entity_id=entity_id,
-                payment_unit=payment_unit,
-                defaults={"entity_name": entity_name},
+        if completed_work is None:
+            completed_work = CompletedWork(
+                opportunity_access=access, entity_id=entity_id, payment_unit=payment_unit, entity_name=entity_name
             )
+            completed_work.save()
+
+    try:
+        with transaction.atomic():
             user_visit.completed_work = completed_work
-            if (
-                counts["daily"] >= payment_unit.max_daily
-                or counts["total"] >= claim_limit.max_visits
-                or (today > claim.end_date or (claim_limit.end_date and today > claim_limit.end_date))
-            ):
-                user_visit.status = VisitValidationStatus.over_limit
-                if not completed_work.status == CompletedWorkStatus.over_limit:
-                    completed_work.status = CompletedWorkStatus.over_limit
-                    completed_work_needs_save = True
-            elif counts["entity"] > 0:
-                user_visit.status = VisitValidationStatus.duplicate
+            user_visit.save()
+    except IntegrityError as e:
+        if "unique_completed_work_deliver_unit_access" not in str(e):
```

Comment:
Probably better to pull this out to a constant `UNIQUE_COMPLETED_WORK_CONSTRAINT='unique_completed_work_deliver_unit_access'`

Reply:
Added. [0e01bad](https://github.com/dimagi/commcare-connect/pull/678/commits/0e01badfbe15b4f2771a3079d5156c22fb8464b9)